### DISCUSSION TITLE
RFC: CLI for generating perfect TS types for XState machines

### DISCUSSION THREAD

**Discussion by mattpocock** [statelyai/xstate#1381] [General] [answered]
## Intro

Static analysis of state machines has enormous potential. In this RFC, I'd like to talk about using **a CLI tool to generate perfect Typescript types** by analysing XState machines in your code.

Typescript with XState is currently imperfect because of XState's innate complexity. The goal is to:

1. Get perfect typings for `MachineOptions` types passed into `Machine`, `interpret` etc. This includes typings of events, services, actions, guards and activities based on their usage in the machine.
2. Get autocomplete on the `matches` function in interpreted state nodes, to allow `state.matches('even.deep.nested.states')` to be type-checkable.

This CLI could, in the future, be tooled to achieve some other **stretch goals**:

1. Autocomplete for states within state machines
2. Analysis of state machine paths so that they error in the IDE, not at runtime
3. Compile only the features of XState you use to save on bundle size.

Running a CLI gives us many, many more options heading into the future.

## Why post the RFC?

**We want to get your thoughts** on how best to implement the types the codegen tool generates, or if there are any alternative approaches we could take - such as VSCode intellisense or others. We are open to any and all ideas.

### 1. Generating sibling files which exports a special 'interpret' function

[Example Output](https://github.com/mattpocock/xstate-codegen/tree/master/packages/sibling-files-with-react-export)

This is how the `xstate-codegen` npm package works currently.

1. Run `xstate-codegen "**/*.machine.ts"` to watch your files.
2. You create a file with a `.machine.ts` extension.
3. The CLI creates a sibling file, with a `.machine.typed.ts` extension.
4. This file creates a custom function to interpret the Machine, depending on configuration passed to the CLI.

> Try it out in this npm package: https://www.npmjs.com/package/xstate-codegen. This is currently tied to React (suited to how my team uses XState) but this could work with any framework.

#### Constraints

The user must use the codegen-generated functions, and pass in the same machine. This is quite a clunky API:

```js
import { interpretLightMachine } from "./trafficLightMachine.machine.typed";
import { lightMachine } from "./trafficLightMachine.machine";

const interpreter = interpretLightMachine(lightMachine, {
  // options
});
```

#### Pro's

Good typings, without overloads.

#### Con's

Clunky API compared with other approaches.

### 2. Global declaration file which overloads XState outputs

[Example Output](https://github.com/mattpocock/xstate-codegen/tree/master/packages/global-env-with-overloads)

1. Run `xstate-codegen "**/*.machine.ts"` to watch your files.
2. The CLI creates a single file, `xstate-codegen-env.d.ts` at the root of your project.
3. For each machine in your project, it creates a `<MachineName>StateMachine` type which can be exported and assigned to your machine:

```ts
import { Machine, TrafficLightStateMachine } from "xstate";

const trafficLightMachine: TrafficLightStateMachine = Machine({
  // machine config
});
```

4. The global `.d.ts` overrides `interpret`, `useMachine` etc so that you can use the machine as you would usually:

```ts
import { interpret } from "xstate";

const service = interpret(trafficLightMachine, {
  // options
});
```

#### User Constraints

The user must pull in and assign the generated type to their machine.

> This could also be automated on file save via a codemod.

#### Pro's

Typings are tied to the machine, which is much better than tying them to the implementation of the machine (i.e. `interpret` etc.). They'll get the compiled types anywhere they use the machine, instead of only in codegen-generated functions.

This approachs asks the least of users - a single import change and you're good to go.

#### Con's

This approach, currently, has the weakest typing. This is because the typings on `interpret` is done by overloads. This renders the typing of machine options next to useless. If the user passes in an incorrect options object, the types fall back to the previous overload, and no error is shown. This is not currently type safe.

> There is more exploratory work to be done on the above to see if any internal changes within XState could be made to facilitate this. I have some ideas which I haven't yet tried.

## 3. Global declaration file without XState Overloads

[Example Output](https://github.com/mattpocock/xstate-codegen/tree/master/packages/global-env-with-compiled-import)

This works the same as 2, but instead of using `interpret` to consume your machine, you would use `interpretCompiled`.

```ts
import { interpretCompiled } from "xstate";

const service = interpretCompiled(trafficLightMachine, {
  // options
});
```

XState would require an update to alias `interpretCompiled` to `interpret`.

### Pro's

Same pro's as 2, with the added benefit of strong typings without any overloads needed.

### Con's

Requires more steps from users to make work.

## 4. @xstate/compiled folder which re-implements xstate

> This has no example output as of yet.

This approach would entirely re-implement the typings of XState in a `node_modules` folder, `@xstate/compiled`. This would allow for maximum customisation of the types by the codegen tool.

It could require the same import syntax as 2, depending on how the internal types are handled:

```ts
import { Machine, TrafficLightStateMachine } from "@xstate/compiled";

const trafficLightMachine: TrafficLightStateMachine = Machine({
  // machine config
});
```

To interpret the machine, users would import from `@xstate/compiled`:

```ts
import { interpret } from "@xstate/compiled";

const service = interpret(trafficLightMachine, {
  // options
});
```

### Pro's

As much control as we like over the typings, without the potential danger of using overloads.

A small change for users to make, which can also be opted into gradually if required.

### Con's

Potentially a more complex implementation for the codegen tool, but this is not necessarily a con for the end user.

**Comment by mattpocock** [comment]
My preferred option is number 4. Implementing an entire reworking of XState's typings give us the most control. When we have full control over all the typings, some really interesting stuff (like the stretch goals mentioned above) become possible.

**Reply by mattpocock** [reply]
Here's where I got to with implementing number 1:

https://twitter.com/mpocock1/status/1288782560621928448

**Comment by montogeek** [comment]
After spending 3 hours typing everything in my first state machine, I very much look forward to this.

**Reply by davidkpiano** [reply]
Hey, it's 3 hours of good typing practice! 😅

**Comment by Andarist** [comment]
Well - we have already discussed this quite a bit, but I'd like to endorse option 4 publicly. The unparalleled control over the types that it gives us is not to be underestimated. 

I somewhat believe that sibling files (implemented for any option) are alluring because working with **giant** files might be a usability problem, but it's also something that could be explored later. Definitely better to stick to the simplest solution and actually test this in real-life scenarios as, right now, this concern is purely hypothetical.

Going back to the point though - it's still escaping me what exact typings we'd like to generate. It feels very tricky to allows any schema while also type-checking it. I currently see only 2 options forward for this:
- insert 2 things with the codegen: returned type AND the schema type
- insert a generated ID into the "inline" schema which would allow TypeScript to match against a discriminated union. This has potential perf limitations though and also assigns a dummy value to the schema (which ultimate is not that big of a deal, just would look slightly weird).

I've prepared [a very simple and dummy TS playground](https://www.typescriptlang.org/play?#code/PTAEHMFMDtIJwIYBdIBNQAcFyQKCQJ4aSgCyCAxgBYCWsAPACoDCA9tCgB5IA0ojAUQBuMXvwDK1SAFsEAPlABeUAG9coDaAD6AZymyAXBP0JcAX1y5UkCgBtsJAGYBXaBSQ12oCnEjJI5NR0kPTqmizsXLxhGoIiHDwxxlQyCEqgCNAEuHIAFHophsmpAJRGgbQMERyQ3HxxovWShfKgloTE-Gw13ADiMOkqoI6srEY6SHB04KAWHSQNHP3Qg-NGAOQCAGoAjOt8WAS2rAioRtDO0gBG8GagAD6qa6CbWwBM+5gIRydnoBNTaDgCz4IgkUgECrBdJQhhJapRZaJTT8YSiJFJNQolETfw6IxY7FEhAEszIonYq4EpIU7HsAQcOAEAD8Rly7k4RgRtSQSNAkCERgE3EQ7iYaKWMD4Kmerw+ZjkJSUCiErBoqBpFIstO1mgscksIFA0HYAFooLBECh0FgcKDOtzuINhqNxpNprN7QsJUhVmCNts9gdvsdTudLjc4HdHjL-S9th9gz8w-93UCLBR2BNQCgJuVIZRKiRlD4-ChYSFHWJFkg8oSNLjc9TaSTVJqiexm7SKYGNld1u3sbqiWTB1S293NPTGQQ2Ry+AKlYoFJnoDpWLZIAA6Y7gXKLwfDz1mEpAA) for us to experiment with this. The goal for me would be to type-check inferred `onEntry`.

There is also a question of how much we'd like to put into the TS compiler itself and how much we could offload to the IDE. The former has the advantage of working in a wider set of scenarios but it would generate much more complex/robust types (maybe it would also come with perf issues at some point? this is just speculation though). OTOH the latter needs to be implemented per IDE - with the VSCode taking a priority from our side as it seems to be the most popular these days.

**Reply by mattpocock** [reply]
Nice, thanks for the TS playground, that'll be useful. 100% agree on the focus on option 4. I'm happy to start exploring option 4 as our default approach, unless any better ideas come from the community.

The IDE approach is interesting, and could definitely improve the experience of editing the code inline. The IDE would be faster at reacting to changes made to the schema inline because it wouldn't require the file being saved, and any autocompleted options would stay as up to date as possible.

I like the idea of keeping a lot of the logic in the TS compiler so that the code generation can run on a CI build. Since option number 4 doesn't generate any actual files in the user's repo, we need to make sure they can get the same experience both in their IDE and in a CI.

> insert a generated ID into the "inline" schema which would allow TypeScript to match against a discriminated union. This has potential perf limitations though and also assigns a dummy value to the schema (which ultimate is not that big of a deal, just would look slightly weird).

Since we have full control over the types, we could make the 'id' or 'key' property required in machines and make that the discriminating key? Then error in the CLI if there are two machines with the same id? That might be a smoother experience for users than having the codegen tool manipulate their code with a codemod.

**Comment by Andarist** [comment]
> Since option number 4 doesn't generate any actual files in the user's repo

Well, technically we could generate global `.d.ts` in the user's repo - like in options 2 and 3. We don't have to write those types in `node_modules`. It seems that maybe in-repo file would have better discoverability and generally would be easier to work with? Not sure about that - Prisma has chosen to write their auto-generated type in `node_modules`, so there's that. I wonder if there are no problems with PnP when doing that 🤔 I've asked about this [here](https://twitter.com/AndaristRake/status/1294186892821397505)

> Then error in the CLI if there are two machines with the same id? That might be a smoother experience for users than having the codegen tool manipulate their code with a codemod.

Not sure how others feel about it but I generally have 0 problems with the tool manipulating my code 😅 I would just treat it like Prettier or ESLint with autofix. I guess the ability to assign the ID manually would be OK, but personally I would much better prefer for it to be auto-generated by default (or at least having an option for this).

A codegen probably wouldn't have a problem with this at all but for the IDE integration we should come up with a strategy to cleanup unused types from the global `.d.ts` (wherever it would live on the fs). 

A conceptual problem that I have with the discriminating key is that TS doesn't actually know what property is used for that - and technically u can have many members in the union with the same value for that key. This means for me that each time TS would have to compare against **all** of the generated schema types to extract those that would be candidates for matching. Maybe this is cheap though? I would expect properties with primitive values to have some kind of precedence for the matching.

**Reply by mattpocock** [reply]
> Well, technically we could generate global .d.ts in the user's repo

If we're getting users to import from `@xstate/compiled`, is there a smart way we can alias that to `xstate` without having to either:

A. Ask them to do a webpack alias
B. Put the .js files in their node_modules?

A is way too clunky. B might be OK? But then it might be confusing if we're putting the .d.ts file in the user's code and the .js files in node_modules.

On the ID stuff, we probably need to do a POC to check it works before we work out the finer details. I'll try and come up with something for the playground to check it works, and isn't too ruinously expensive.

**Reply by Andarist** [reply]
The `@xstate/compiled` package could just reexport `xstate` - as in: it could be a real package. We’d have to figure out how to make it not-typed, so it wouldnt reexport `xstate` types. Or maybe a local .d.ts just would take a precedence? Not sure how types are resolved exactly and when there is a potential for conflicts and when they are simply overriden.

**Reply by mattpocock** [reply]
Had a quick check on the above, and it works really nicely - you can export the js files, but not the types. You can then either declare the types with a local .d.ts file or inside node_modules.

**Reply by Andarist** [reply]
Right - just have thought about it 😅 had a brainfart before

**Reply by mattpocock** [reply]
Have made some nice progress with the 'id' idea.

```ts
type LightEvent =
  | { type: 'TIMER' }
  | { type: 'POWER_OUTAGE' }
  | { type: 'PED_COUNTDOWN'; duration: number };

interface LightContext {
  elapsed: number;
}

// id is passed in as a third argument here
const lightMachine = Machine<LightContext, LightEvent, 'lightMachine'>({
  // config
});

```

Passing in an id as a generic means we can reference it in an interface map:

```ts
export interface RegisteredMachinesMap<
  TContext,
  TStateSchema,
  TEvent extends EventObject
> {
  lightMachine: LightMachineStateMachine<
    TContext,
    TStateSchema,
    TEvent,
    'lightMachine'
  >;
  faceMachine: FaceMachineStateMachine<
    TContext,
    TStateSchema,
    TEvent,
    'faceMachine'
  >;
}
```

This is better than this implementation:

```ts
import { TrafficLightMachine, Machine } from '@xstate/compiled';

const lightMachine: TrafficLightMachine = Machine<LightContext, LightEvent>({
  // config
});
```

Because the generics don't need to be copied over to our `.d.ts` file. The generics can be passed all the way through from the user's code to the `interpret` function without needing to be duplicated.

I'll tidy up my code and push something.

**Reply by mattpocock** [reply]
Here's the example implementation. This is by no means complete because it doesn't re-export all of XState's other typings, but is a POC for the idea above.

https://github.com/mattpocock/xstate-codegen/tree/master/packages/xstate-compiled

**Reply by mattpocock** [reply]
@Andarist Have made a breakthrough with typing entry actions based on their event sources _and_ autocompleting targets.

https://twitter.com/mpocock1/status/1294311443941265414

**Reply by mattpocock** [reply]
More detail: https://www.youtube.com/watch?v=5hCAQSmIpy4

**Reply by CodingDive** [reply]
> We don't have to write those types in `node_modules`. It seems that maybe in-repo file would have better discoverability and generally would be easier to work with? Not sure about that - Prisma has chosen to write their auto-generated type in `node_modules`, so there's that.

I just remembered our conversation about stateschemas. To get handy extra documentation with all the amazing typegeneration, you indeed don't want to look in your node_modules folder. I'd too much rather have the types output in the same folder as the machine.

From the stateschema issue https://github.com/davidkpiano/xstate/issues/943, I wrote:
> yes, good visualization or autogeneration would definitely be a very good solution to my problem [of having to define typestates instead of a stateschema]

**Reply by mattpocock** [reply]
@CodingDive Just saw this. I think that the output might not be great for documentation because it's pretty verbose and mangled. You can try running latest here:

https://github.com/mattpocock/xstate-codegen/tree/master/packages/xstate-compiled

**Reply by mattpocock** [reply]
@CodingDive We'll be adding a custom outDir for the type declarations so you can commit them if you like.

**Comment by jamesopstad** [comment]
This looks really interesting. I was wondering recently whether `xstate` could be implemented as a compiler (in a similar way to Svelte) and this seems like a step in that direction. The thing that bothers me about the approaches listed above is tying this to the React `useMachine` hook. One of the things I like most about `xstate` is that it is framework independent. Could the types not be generated purely based on the machine definition? Additional options passed to `useMachine` would then be checked against these imported types. I'm sure there are good reasons why this wouldn't work but it would be good to understand more.

**Reply by mattpocock** [reply]
@jamesopstad 100%, I totally agree with this direction. Adopting 4 means we can generate types tied to the state machine, which can then be consumed by whichever interpreter you choose (useMachine, interpret etc.)

**Reply by davidkpiano** [reply]
> The thing that bothers me about the approaches listed above is tying this to the React

I would agree with this. @mattpocock Would it be possible to update the RFC so that it wasn't tied to any framework/library except for `xstate`? 🙏

**Reply by mattpocock** [reply]
@davidkpiano Updated to remove references to useMachine to make it more closely tied to XState.

**Reply by bwestergard** [reply]
@davidkpiano Thank you for all your hard work on this project.

Building on @jamesopstad's broader suggestion: Do you think it would ever make sense to eliminate the `Machine` and `createMachine` functions in favor of a compile step that transformed the JSON machine definition to a fully statically typed machine implementation?

**Reply by Andarist** [reply]
That would be possible but also it wouldn't bring us many benefits - I mean, we can already do this using `Machine`/`createMachine` APIs and those are great pointers for any processing tool as to which objects should be processed by them and which don't.

**Reply by bwestergard** [reply]
@Andarist Hm, I'm not sure I understand. Would using a compilation step to produce something code [along these lines](https://github.com/davidkpiano/xstate/discussions/1975#discussioncomment-491288) ensure that `Context` in different states can be fully statically checked, which doesn't seem to be possible within Typescript right now?

**Reply by Andarist** [reply]
It would be great to use discriminated unions somehow but this might be quite challenging. I'm not yet sure how we are going to pass around a particular `Context` subtype so TS can understand it. We'll be trying to figure this out in the future. When? Hard to tell.

**Comment by alavkx** [comment]
I haven't any strong opinion. I only want to thank you—this is really exciting stuff :)

**Reply by mattpocock** [reply]
Thank you so much! I'm excited too ❤️

**Comment by fli** [comment]
https://www.pulumi.com/blog/introducing-crd2pulumi/

Something with a similar concept I used today which went pretty smoothly.
In this case you run the CLI tool on some yaml which represents a resource, then it creates a typed version of the resource in a new folder which you can just import. Imported resource has all typescript typings.

Just some prior art/reference which might be useful to you all :)

**Comment by mattpocock** [comment, marked as answer]
OK, we've released our first RC version which implements number 4 of the above:

https://www.npmjs.com/package/xstate-codegen/v/next

Here we go, folks! Time to start testing and seeing what we've missed. It's still POC, not production-ready yet. Looking forward to your thoughts.

**Reply by alavkx** [reply]
Incredible effort!

**Comment by mattpocock** [comment]
Getting closer to merging the 'next' branch into master. Want to help out? Download the 'next' branch from npm and help us bugfix in your repo:

https://www.npmjs.com/package/xstate-codegen/v/next
